osf/models/user.py

```diff
@@ -1508,6 +1508,56 @@ def get_node_comment_timestamps(self, target_id):
         default_timestamp = dt.datetime(1970, 1, 1, 12, 0, 0, tzinfo=pytz.utc)
         return self.comments_viewed_timestamp.get(target_id, default_timestamp)
 
+    def gdpr_delete(self):
+        """
+        This function does not remove the user object reference from our database, but it does disable the account and
+        remove identifying in a manner compliant with GDPR guidelines.
+        :return:
+        """
+        from osf.models import PreprintService, AbstractNode
+
+        user_nodes = self.nodes.all()
+        #  Validates the user isn't trying to delete things they deliberately made public.
+        if user_nodes.filter(type='osf.registration').exists():
+            raise ValidationError('You cannot delete this user because they have one or more registrations.')
+
+        if PreprintService.objects.filter(node___contributors=self, ever_public=True).exists():
+            raise ValidationError('You cannot delete this user because they have one or more preprints.')
+
+        # Validates that the user isn't trying to delete things nodes they are the only admin on.
+        personal_nodes = [node.id for node in user_nodes if node.contributor_set.count() == 1]
+        personal_nodes = AbstractNode.objects.filter(id__in=personal_nodes)
+        shared_nodes = user_nodes.exclude(id__in=personal_nodes.values_list('id'))
+
+        for node in shared_nodes.exclude(type='osf.quickfilesnode'):
+            alternate_admins = Contributor.objects.select_related('user').filter(
+                node=node,
+                user__is_active=True,
+                admin=True,
+            ).exclude(user=self)
+            if not alternate_admins:
+                raise ValidationError(
+                    'You cannot delete node {} because it would be a node with contributors, but with no admin.'.format(
+                        node._id))
+
+        # delete all personal nodes (one contributor), bookmarks, quickfiles etc.
+        personal_nodes.delete()
+
+        # This removes identifying info
+        self.fullname = 'Deleted User'
+        self.username = 'Deleted User'
+        self.given_name = 'Deleted User'
```

Comment:
Drive-by: I would suggest setting `fullname` to "Deleted User" and setting `given`, `family`, and `middle` to an empty string. There may be place where we display names as given + family which would lead to "Deleted User Deleted User" in this case.

Reply:
And let's clear `suffix` while we're at it.

Reply:
I haven't looked into these two fields to see how they're used, but I would think we should clear `mailchimp_mailing_lists` and `osf_mailing_lists` as well so that a mailing list doesn't end up with a null email.